### PR TITLE
Fix ignoring the exclude in the logger's record function for json, csv and log logging formats

### LINKED ISSUE #187: [Bug] Excluding json or csv formats is ignored by the logger's record function
**Describe the bug**
When running `record("some_log_tag", some_data, exclude=("json", "csv"))` the `exclude` parameter is ignored and `some_data` is still logged to json and csv.

**Code example**
The following is a minimal test code snipped that reproduces the bug and could also be used as a basis for an automated regression test

```python
from stable_baselines3.common.logger import make_output_format, read_json


def test_ignore_json_formatter(tmp_path):
    writer = make_output_format("json", tmp_path)
    writer.write(dict(some_tag=42), dict(some_tag=("json")))
    assert read_json(tmp_path / "progress.json").empty
```
For a real regression test I'd use test parametrization to check all the different format writers in one test, following the pattern of `test_logger.py`.

**System Info**
Describe the characteristic of your environment:
 * Install: pip install -e .[extra] within the source root directory
 * python 3.8
 * PyTorch 1.6.0
 * Gym 0.17.3

As already mentioned in https://github.com/DLR-RM/stable-baselines3/issues/186 I'd be happy to provide a PR with a fix and regression tests. Let me know if there is anything else I should consider, or if I should proceed with a PR.

### PR DESCRIPTION
**Pull request by SwamyDev**
## Description
I've introduced a general `filter_excluded_keys` function which takes `key_values` and `key_excluded` dictionaries as well as a logging `_format` specifier to remove excluded keys from the  `key_values` dictionary. This function is then used in `JSONOutputFormat` and `CSVOutputFormat` to remove excluded keys before any further processing is done. In `HumanOutputFormat` I've simply added an additional check for the `log` format type. 

In `test_logger.py` I've introduced regression tests, that check that all `*Format` objects produce no output when the key written is excluded. To do that I've introduced a new `read_log` fixture, which reads the contents from files, captured standard output and Tensorboard events and wraps it in a content interface. The wrapper makes the test check more convenient because I can just assert for the empty property for any logging format type. I've also used the `read_log` fixture in the `test_make_output` test, to make sure that the read_log function is actually reading content when it should. As a side effect the `test_make_output` function is imo a bit more thorough by also checking for content of log and Tensorboard.

## Motivation and Context
Currently using the `exclude` parameter of the logger's `record` function would be ignored for csv, json and log formats. 
closes #187 

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist:
- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)

## Side Notes
I've also been running `make commit-checks` and `make type`, however I do get to errors in `base_class.py` even on a fresh clone:
```bash
File "/tmp/stable-baselines3/stable_baselines3/common/base_class.py", line 587, in load: Can't instantiate BaseAlgorithm with abstract methods _setup_model, learn [not-instantiable]
File "/tmp/stable-baselines3/stable_baselines3/common/base_class.py", line 587, in load: Invalid keyword argument _init_setup_model to function BaseAlgorithm.__init__ [wrong-keyword-args]
         Expected: (self, policy, env, policy_base, learning_rate, policy_kwargs, tensorboard_log, verbose, device, support_multi_env, create_eval_env, monitor_wrapper, seed, use_sde, sde_sample_freq)
  Actually passed: (self, policy, env, device, _init_setup_model)
```
Is this to be expected as it is still part of the rework, or is this possibly a problem with my pytype installation (version 2020.10.8)?

### CODE REVIEW

**Comment by araffin** [comment]
>Is this to be expected as it is still part of the rework, or is this possibly a problem with my pytype installation (version 2020.10.8)?

That's weird, I don't have any error (Ubuntu 18.04, python 3.6, same pytype version) and github CI passes too...

File: stable_baselines3/common/logger.py [hunk 1]
```diff
@@ -140,6 +140,23 @@ def close(self) -> None:
...
+    key_values: Dict[str, Any], key_excluded: Dict[str, Union[str, Tuple[str, ...]]], _format: str
+) -> Dict[str, Any]:
+    """
+    filters the keys specified by `key_exclude` for the specified format
+
+    :param key_values: log dictionary to be filtered
+    :param key_excluded: keys to be excluded per format
+    :param _format: format for which this filter is run
+    """
+
```
Comment (araffin) [inline stable_baselines3/common/logger.py:153]:
```suggestion
    """
    Filters the keys specified by ``key_exclude`` for the specified format

    :param key_values: log dictionary to be filtered
    :param key_excluded: keys to be excluded per format
    :param _format: format for which this filter is run
    :return: dict without the excluded keys
    """

```

File: stable_baselines3/common/logger.py [hunk 2]
```diff
@@ -140,6 +140,23 @@ def close(self) -> None:
...
+    def is_excluded(key):
```
Comment (araffin) [inline stable_baselines3/common/logger.py:154]: missing types

File: stable_baselines3/common/logger.py [hunk 3]
```diff
@@ -150,18 +167,20 @@ def __init__(self, filename: str):
         self.file = open(filename, "wt")
 
     def write(self, key_values: Dict[str, Any], key_excluded: Dict[str, Union[str, Tuple[str, ...]]], step: int = 0) -> None:
-        for (key, value), (_, excluded) in zip(sorted(key_values.items()), sorted(key_excluded.items())):
-
-            if excluded is not None and "json" in excluded:
-                continue
-
+        def cast_to_json_serializable(value):
```
Comment (araffin) [inline stable_baselines3/common/logger.py:170]: missing types

**Review by araffin — CHANGES_REQUESTED** [review]
Thanks for the good PR =)
only minor changes needed and we are good to merge ;)

File: stable_baselines3/common/logger.py (hunk 2 above)
Comment (SwamyDev) [reply stable_baselines3/common/logger.py]: fixed

**Comment by SwamyDev** [comment]
Thanks glad you liked it :) I've applied the fixes requested

As for pytype it seems that the python version makes the difference. I get the error with python 3.8.5, but with 3.6 it's no problem. I'm running Ubuntu 20.04 and I've just set up a simple virtualenv before. With conda and python 3.6 it does work however.

File: stable_baselines3/common/logger.py [hunk 4]
```diff
@@ -140,6 +140,24 @@ def close(self) -> None:
...
+    """
+    Filters the keys specified by ``key_exclude`` for the specified format
+
+    :param key_values: log dictionary to be filtered
+    :param key_excluded: keys to be excluded per format
+    :param _format: format for which this filter is run
+    :return: dict without the excluded keys
+    """
+
+    def is_excluded(key: str):
```
Comment (araffin) [inline stable_baselines3/common/logger.py:155]:
```suggestion
    def is_excluded(key: str) -> bool:
```

Comment (SwamyDev) [reply stable_baselines3/common/logger.py]: Darn, sry that one slipped through again - not used to these python type hints ;)

**Review by araffin — APPROVED** [review]
LGTM =)

**Pull request merged**